editoast/src/views/test_app.rs

```diff
@@ -171,12 +206,95 @@ where
     }
 }
 
-impl<S, Sref> AsRef<Sref> for TestApp<S>
+impl<S> AsRef<S> for TestApp<S>
 where
     S: Service<Request, Response = ServiceResponse<BoxBody>, Error = Error>,
-    S: AsRef<Sref>,
 {
-    fn as_ref(&self) -> &Sref {
-        self.deref().as_ref()
+    fn as_ref(&self) -> &S {
+        &self.service
+    }
+}
+
+pub struct TestResponse {
+    inner: ServiceResponse<BoxBody>,
+    log_payload: bool,
+}
+
+impl TestResponse {
+    fn new(inner: ServiceResponse<BoxBody>) -> Self {
+        tracing::debug!(target: "test_app", response = ?inner, "Recieved test response");
+        Self {
+            inner,
+            log_payload: true,
+        }
+    }
+
+    #[allow(unused)]
+    pub fn log_payload(mut self, log_payload: bool) -> Self {
+        self.log_payload = log_payload;
+        self
+    }
+
+    fn render_response_lossy(self) -> String {
+        if !self.log_payload {
+            return "payload logging disabled".to_string();
+        }
+        self.inner
+            .into_body()
+            .try_into_bytes()
+            .ok()
+            .and_then(|bytes| serde_json::from_slice::<serde_json::Value>(&bytes).ok())
+            .and_then(|json| serde_json::to_string_pretty(&json).ok())
+            .unwrap_or_else(|| "cannot render response body".to_string())
+    }
+
+    pub fn assert_status(self, expected_status: actix_http::StatusCode) -> Self {
+        let actual_status = self.inner.status();
+        if actual_status != expected_status {
+            let body = self.render_response_lossy();
+            tracing::error!(
+                target: "test_app",
+                ?expected_status,
+                ?actual_status,
+                %body,
+                "Unexpected status code"
+            );
+            pretty_assertions::assert_eq!(actual_status, expected_status, "unexpected status code");
+            unreachable!("should have already panicked")
+        } else {
+            self
+        }
+    }
```

Comment:
How about structuring the function differently to avoid this `unreachable!()` hack (because I suppose this is only to please a warning of the compiler, right?).

```rust
let actual_status = self.inner.status();
let body = self.render_response_lossy();
pretty_assertions::assert_eq!(actual_status, expected_status, "unexpected status code body={:?}", body);
self
```

That means not using `tracing`, but since in tests, `tracing` is only configured to output on `stdout` and `assert_eq!()` will also output on `stdout`, I believe it's fine, what do you think?

Reply:
The problem is that to read the response payload we need to consume the response object, so we can't `return self` afterwards.

I merged the `error!` into the `assert_eq!` still.